src/Heading/Heading.tsx

```diff
@@ -0,0 +1,81 @@
+import { ElementType, FC } from 'react';
+import { SetOptional } from 'type-fest';
+import styled from '@emotion/styled';
+import { css, Theme } from '@emotion/react';
+
+import { FontSize, FontWeight } from '../theme/typography';
+import Text from '../Text';
+
+const lineHeightMultiplier = 0.5;
+
+const baseStyles = ({ theme }: { theme: Theme }) => css`
+  line-height: ${theme.lineHeights.small};
+`;
+
+const calculatedMargin = ({
+  fontSize,
+  mb,
+  theme
+}: {
+  fontSize: FontSize;
+  mb: string;
+  theme: Theme;
+}) =>
+  !mb &&
+  css`
+    margin-bottom: calc(${theme.fontSizes[fontSize]} * ${lineHeightMultiplier});
+  `;
+
+type Props = {
+  fontSize: FontSize;
+  fontWeight?: FontWeight;
+  mb: string;
+};
+
+const Heading = styled(Text)<Props>`
+  ${baseStyles};
+  ${calculatedMargin};
+`;
+
+Heading.defaultProps = {
+  as: 'h1',
+  color: 'text.dark',
+  fontWeight: 'bold',
+  lineHeight: 'small'
+};
+
+const createHeading =
+  (
+    as: ElementType,
+    defaultFontSize: FontSize
+  ): FC<SetOptional<Props, 'fontSize'>> =>
```

Comment:
https://github.com/sindresorhus/type-fest/blob/main/source/set-optional.d.ts

`fontSize` is required for `Heading`, but `H1`, `H2`, etc. have default `fontSizes`, so it's not required for those.